File: parking-lot/go-without-multithreading/entities/level.go
package entities
import (
    "app/parking-lot/go-without-multithreading/enums"
    "fmt"
)

type Level struct {
	ID             int
	Spots          []*ParkingSpot
	AvailableSpots map[enums.VehicleType]int
}

func NewLevel(id int, numMotorcycleSpots, numCarSpots, numTruckSpots int) *Level {
	level := &Level{
		ID:             id,
		Spots:          make([]*ParkingSpot, 0, numMotorcycleSpots+numCarSpots+numTruckSpots),
		AvailableSpots: make(map[enums.VehicleType]int),
	}

	level.AvailableSpots[enums.Motorcycle] = numMotorcycleSpots
	level.AvailableSpots[enums.Car] = numCarSpots
	level.AvailableSpots[enums.Truck] = numTruckSpots

	spotNum := 1
	for i := 0; i < numMotorcycleSpots; i++ {
		level.Spots = append(level.Spots, &ParkingSpot{ ID: fmt.Sprintf("L%d-M%d", id, spotNum), Type: enums.Motorcycle, LevelID: id })
		spotNum++
	}
	spotNum = 1
	for i := 0; i < numCarSpots; i++ {
		level.Spots = append(level.Spots, &ParkingSpot{ ID: fmt.Sprintf("L%d-C%d", id, spotNum), Type: enums.Car, LevelID: id })
		spotNum++
	}
	spotNum = 1
	for i := 0; i < numTruckSpots; i++ {
		level.Spots = append(level.Spots, &ParkingSpot{ ID: fmt.Sprintf("L%d-T%d", id, spotNum), Type: enums.Truck, LevelID: id })
		spotNum++
	}

	return level
}

// Simple modifier decoupled from execution math
func (l *Level) ModifyAvailability(vType enums.VehicleType, delta int) {
    l.AvailableSpots[vType] += delta
}

File: parking-lot/java-without-multithreading/entities/Level.java
package entities;
import enums.VehicleType;
import java.util.*;

public class Level {
    private int id;
    private List<ParkingSpot> spots;
    private Map<VehicleType, Integer> availableSpots;

    public Level(int id, int numMotorcycleSpots, int numCarSpots, int numTruckSpots) {
        this.id = id;
        this.spots = new ArrayList<>();
        this.availableSpots = new HashMap<>();

        availableSpots.put(VehicleType.MOTORCYCLE, numMotorcycleSpots);
        availableSpots.put(VehicleType.CAR, numCarSpots);
        availableSpots.put(VehicleType.TRUCK, numTruckSpots);

        int spotNum = 1;
        for(int i=0; i<numMotorcycleSpots; i++) { spots.add(new ParkingSpot(String.format("L%d-M%d", id, spotNum++), VehicleType.MOTORCYCLE)); }
        spotNum = 1;
        for(int i=0; i<numCarSpots; i++) { spots.add(new ParkingSpot(String.format("L%d-C%d", id, spotNum++), VehicleType.CAR)); }
        spotNum = 1;
        for(int i=0; i<numTruckSpots; i++) { spots.add(new ParkingSpot(String.format("L%d-T%d", id, spotNum++), VehicleType.TRUCK)); }
    }

    public int getId() { return id; }
    public List<ParkingSpot> getSpots() { return spots; }
    public Map<VehicleType, Integer> getAvailableSpots() { return availableSpots; }

    public void modifyAvailability(VehicleType type, int delta) {
        availableSpots.put(type, availableSpots.get(type) + delta);
    }
}
